spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java

```diff
@@ -0,0 +1,126 @@
+package org.springframework.cloud.consul.discovery;
+
+import com.ecwid.consul.v1.ConsulClient;
+import com.ecwid.consul.v1.agent.model.NewService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+
+import javax.annotation.PreDestroy;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Created by nicu on 11.03.2015.
+ */
+@Slf4j
+public class TtlScheduler implements ApplicationContextAware {
+    private static final AtomicInteger _instances = new AtomicInteger();
+    private static final int DEFAULT_TTL = 3; // must be > 1
+    public static final int HEARTBEAT_INTERVAL_RATIO = 2 / 3;
+
+    private final ScheduledExecutorService ttlExecutor;
+    private volatile ScheduledFuture<?> scheduledFuture;
+    private final Thread shutdownThread;
+    private final Set<String> serviceIds;
+    private final AtomicBoolean isShuttingDown;
+    private final Runnable ttlPingThread;
+    private volatile int ttl;
+    private volatile int heartbeatInterval;
+
+    @Autowired
+    private ConsulClient client;
+
+    public TtlScheduler() {
+        if (_instances.addAndGet(1) > 1) {
+            throw new IllegalStateException("Expecting this to be used in singleton mode!");
+        }
+        //one thread to refresh ttl for each registered app to local agent is ok
+        ttlExecutor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Spring has great facilities for handling scheduled executions: `@Scheduled`.  It handles startup, thread management and shutdown.